### Import Libraries

In [1]:
from helper import *
from main import *
import pandas as pd
from sklearn.svm import SVC
from sklearn import svm

### Config

In [2]:
# Data path
data_path = '/Users/gabrielvictorgomesferreira/Library/Mobile Documents/com~apple~CloudDocs/Work/ISU Classes/COMS 573 - Machine Learning/data/data_hw2/'
train_file = 'train.txt'
test_file = 'test.txt'

### Load, Extract two-dimensional feature (symmetry and average intensity) and Split DFs

In [3]:
X_train, y_train = load_features(data_path + train_file)
X_test, y_test = load_features(data_path + test_file)

print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(1561, 2) (1561,)
(424, 2) (424,)


In [4]:
def svm_with_diff_c(train_label, train_data, test_label, test_data):
    '''
    Use different value of cost c to train a svm model. Then apply the trained model
    on testing label and data.
    
    The value of cost c you need to try is listing as follow:
    c = [0.01, 0.1, 1, 2, 3, 5]
    Please set kernel to 'linear' and keep other parameter options as default.
    No return value is needed
    ''' 

    ### YOUR CODE HERE
    print(f"{'-'*95}")
    print("Using different value of cost c to train a svm model. Then apply the trained model on testing label and data.")
    print("")
    # List of Cs to try
    c_values = [0.01, 0.1, 1, 5, 10]
    
    # Loop through each value of C and train an SVM model
    for c in c_values:

        # Build SVM model with linear kernel and cost parameter C
        model = svm.SVC(C=c, kernel='linear')
        
        # Train the model
        model.fit(train_data, train_label)
        
        # Calculate accuracy on the test data
        accuracy = model.score(test_data, test_label)
        
        # Get the number of support vectors
        num_support_vectors = len(model.support_)
        
        # Print the results
        print(f"C={c} | Accuracy={accuracy:.4f} | Support Vectors={num_support_vectors}")
        print("")

    print(f"{'-'*95}")
    print(" ")

    ### END YOUR CODE
    

def svm_with_diff_kernel(train_label, train_data, test_label, test_data):
    '''
    Use different kernel to train a svm model. Then apply the trained model
    on testing label and data.
    
    The kernel you need to try is listing as follow:
    'linear': linear kernel
    'poly': polynomial kernel
    'rbf': radial basis function kernel
    Please keep other parameter options as default.
    No return value is needed
    '''

    ### YOUR CODE HERE
    print(f"{'-'*95}")
    print("Using different kernel funcitons to train a svm model. Then apply the trained model on testing label and data.")
    print("")
    # List of kernels to try
    kernels = ['linear', 'poly', 'rbf']
    
    # Loop through each kernel type and train an SVM model
    for kernel in kernels:

        # Create the SVM model with the specified kernel
        model = svm.SVC(kernel=kernel)
        
        # Train the model
        model.fit(train_data, train_label)
        
        # Calculate accuracy on the test data
        accuracy = model.score(test_data, test_label)
        
        # Get the number of support vectors
        num_support_vectors = len(model.support_)
        
        # Print the results
        print(f"Kernel={kernel} | Accuracy={accuracy:.4f} | Support Vectors={num_support_vectors}")
        print("")
    print(f"{'-'*95}")
    ### END YOUR CODE

In [5]:
svm_with_diff_c(y_train, X_train, y_test, X_test)

-----------------------------------------------------------------------------------------------
Using different value of cost c to train a svm model. Then apply the trained model on testing label and data.

C=0.01 | Accuracy=0.8939 | Support Vectors=1080

C=0.1 | Accuracy=0.9623 | Support Vectors=414

C=1 | Accuracy=0.9599 | Support Vectors=162

C=5 | Accuracy=0.9623 | Support Vectors=104

C=10 | Accuracy=0.9623 | Support Vectors=92

-----------------------------------------------------------------------------------------------
 


* *C = 0.01:* Achieved an accuracy of 89.39% using 1080 support vectors, indicating underfitting due to the large margin.

* *C = 0.1:* Accuracy significantly improved to 96.23% with 414 support vectors. This shows that increasing C reduces the margin width, resulting in better performance.

* *C = 1:* Accuracy slightly decreased to 95.99%, with the number of support vectors dropping further to 162, indicating that the decision boundary is becoming more precise.

* *C = 5:* Accuracy returned to 96.23% using only 104 support vectors, demonstrating that the model is reaching optimal performance with less support vectors as C increases.

* *C = 10:* Accuracy remained stable at 96.23% with 92 support vectors, suggesting that further increases in C have a minimal impact on performance, but lead to a titgh margin and with less support vectors.

In [6]:
svm_with_diff_kernel(y_train, X_train, y_test, X_test)

-----------------------------------------------------------------------------------------------
Using different kernel funcitons to train a svm model. Then apply the trained model on testing label and data.

Kernel=linear | Accuracy=0.9599 | Support Vectors=162

Kernel=poly | Accuracy=0.9575 | Support Vectors=75

Kernel=rbf | Accuracy=0.9623 | Support Vectors=90

-----------------------------------------------------------------------------------------------


* *Linear Kernel:* Achieved an accuracy of 95.99% using 162 support vectors. The model performed well for linearly separable data but required a relatively high number of support vectors.

* *Polynomial Kernel:* Accuracy slightly decreased to 95.75%, while the number of support vectors dropped significantly to 75.

* *RBF Kernel:* Achieved the highest accuracy of 0.9623 with only 90 support vectors, indicating that the RBF kernel effectively handles non-linear boundaries with a good balance between accuracy and support vector count.